xmtp_mls/src/storage/encrypted_store/mod.rs

```diff
@@ -163,8 +165,155 @@ fn warn_length<T>(list: &Vec<T>, str_id: &str, max_length: usize) {
+    type Key = Vec<u8>;
+    fn fetch(&mut self, key: Vec<u8>) -> Result<Option<StoredKeyStoreEntry>, StorageError> where {
+        use self::schema::openmls_key_store::dsl::*;
+        Ok(openmls_key_store.find(key).first(self).optional()?)
+    }
+}
+
+impl Delete<StoredKeyStoreEntry> for DbConnection {
+    type Key = Vec<u8>;
+    fn delete(&mut self, key: Vec<u8>) -> Result<usize, StorageError> where {
+        use self::schema::openmls_key_store::dsl::*;
+        Ok(diesel::delete(openmls_key_store.filter(key_bytes.eq(key))).execute(self)?)
+    }
+}
+
+impl Store<DbConnection> for StoredIdentity {
+    fn store(&self, into: &mut DbConnection) -> Result<(), StorageError> {
+        diesel::insert_into(identity::table)
+            .values(self)
+            .execute(into)?;
+        Ok(())
+    }
+}
+
+impl Fetch<StoredIdentity> for DbConnection {
+    type Key = ();
+    fn fetch(&mut self, _key: ()) -> Result<Option<StoredIdentity>, StorageError> where {
+        use self::schema::identity::dsl::*;
+        Ok(identity.first(self).optional()?)
+    }
+}
+
 #[cfg(test)]
 mod tests {
+    use super::{models::*, EncryptedMessageStore, StorageError, StorageOption};
+    use crate::{Fetch, Store};
+    use rand::{
+        distributions::{Alphanumeric, DistString},
+        Rng,
+    };
+    use std::boxed::Box;
+    use std::fs;
+
+    fn rand_string() -> String {
+        Alphanumeric.sample_string(&mut rand::thread_rng(), 16)
+    }
+
+    fn rand_vec() -> Vec<u8> {
+        rand::thread_rng().gen::<[u8; 16]>().to_vec()
+    }
+
+    #[test]
+    fn ephemeral_store() {
+        let store = EncryptedMessageStore::new(
+            StorageOption::Ephemeral,
+            EncryptedMessageStore::generate_enc_key(),
+        )
+        .unwrap();
+        let conn = &mut store.conn().unwrap();
+
+        let account_address = "address";
+        StoredIdentity::new(account_address.to_string(), rand_vec(), rand_vec())
+            .store(conn)
+            .unwrap();
+
+        let fetched_identity: StoredIdentity = conn.fetch(()).unwrap().unwrap();
+        assert_eq!(fetched_identity.account_address, account_address);
+    }
+
+    #[test]
+    fn persistent_store() {
+        let db_path = format!("{}.db3", rand_string());
+        {
+            let store = EncryptedMessageStore::new(
+                StorageOption::Persistent(db_path.clone()),
+                EncryptedMessageStore::generate_enc_key(),
+            )
+            .unwrap();
+            let conn = &mut store.conn().unwrap();
+
+            let account_address = "address";
+            StoredIdentity::new(account_address.to_string(), rand_vec(), rand_vec())
+                .store(conn)
+                .unwrap();
+
+            let fetched_identity: StoredIdentity = conn.fetch(()).unwrap().unwrap();
+            assert_eq!(fetched_identity.account_address, account_address);
+        }
+
+        fs::remove_file(db_path).unwrap();
```

Comment:
Not a blocker, but we should probably have some sort of `defer` like struct that lets us run cleanup tasks like this even if the code above it errors, which should run any time the struct gets dropped from memory. We could do it with something like this:

```rust
struct FileGuard {
    path: String,
}

impl FileGuard {
    fn new(path: String) -> Self {
        FileGuard { path }
    }
}

impl Drop for FileGuard {
    fn drop(&mut self) {
        if let Err(e) = fs::remove_file(&self.path) {
            eprintln!("Failed to remove {}: {}", self.path, e);
        }
    }
}
```

Reply:
Or we could use a tempfile. I just know that I have a bunch of random `.db3` files sitting around in my local copy of this repo

Reply:
+1 for tempfile